cumulusci/tasks/metadata_etl/tests/test_base.py

```diff
@@ -379,3 +381,173 @@ def test_transform__xml_parse_error(self):
             test_path.write_text(">>>>>NOT XML<<<<<")
             with pytest.raises(etree.ParseError):
                 task._transform()
+
+
+class TestUpdateMetadataFirstChildTextTask:
+    def test_init_options(self):
+        for options, expected_value in [
```

Comment:
This test style is very hard to read (and hard to evaluate when it fails). Please break out these cases into separate unit tests. A utility method or Pytest fixture can help elide the boilerplate while preserving the clarity and independence of each test case.

Reply:
Working on making tests readable.   Looking into [https://docs.pytest.org/en/stable/parametrize.html](https://docs.pytest.org/en/stable/parametrize.html)

Reply:
If I were writing the tests I might do them like this:

    def test_init_options__no_namespace(self):
        task = create_task(UpdateMetadataFirstChildTextTask, {
            "managed": False,
            "namespace_inject": None,
            "entity": "CustomObject",
            "tag": "customObjectAttribute",
            "value": "newAttributeValue",
        })
        assert options["entity"] == task.entity
        assert task.options["value"] == "newAttributeValue"

with a separate function for each test case, and not worry about the repeated assertions. This way there's a hint in the function name about what scenario is being tested. Tests are _usually_ not really the place to abstract things.

That said, I think using pytest parametrize is an acceptable way to go here too (it at least gives more helpful output when one of the sets of parameters fails) and you may use it if you prefer.

Reply:
@davidmreed  and @davisagli 

I split up `test_base.py`'s `TestUpdateMetadataFirstChildTextTask` tests:  one for each expected outcome using `@pytest.mark.parametrize`.   Thoughts?   I can split out my scenarios into separate tests too.

Question:  My difficulty is testing `_init_options`: `_init_options` is called during the task's `__init__`.   It appears I can only test `_init_options` with functional tests.   Is this correct?   A unit test would be much shorter.

Reply:
@spelak-salesforce I think separate tests would be easier to read, but I also think this is good enough and not worth spending more time on. I did make a commit to consolidate the two tests for _init_options since they had exactly the same test code.

You're right that _init_options can't be called without instantiating the task, which runs it implicitly via `__init__`. I would still say this test is closer to the unit test end of the spectrum than functional since it's testing that a portion of the code works properly rather than testing the functionality seen by the user. I suppose the framework could be changed so that option parsing happens in pure functions that get attached to task classes. But I don't think it's worth the backwards incompatible change in the framework unless this current approach becomes a performance bottleneck. Anyway we have some other changes in mind for task options that will make it possible for them to be specified declaratively (and thus not need special code in _init_options) in many more cases...